### PR TITLE
GaussianMixtureModel

### PR DESCRIPTION
**Pull request by KotaroSetoyama**
Created GMM prototype. This prototype measures the training and evaluation speed of GMM on CPU and GPU. To compare the speed with scikit-learn, I referenced the sklearn code for the most part. Although this current version probably has some redundant parts and has to be recreated later, you can confirm the the speed of GPU.

### CODE REVIEW

File: examples/gmm/gmm.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        log_prob_norm = self.xp.log(self.xp.sum(self.xp.exp(weighted_log_prob),
+                                                axis=1))
+        with np.errstate(under='ignore'):
+            log_resp = weighted_log_prob - log_prob_norm[:, self.xp.newaxis]
+
+        return self.xp.mean(log_prob_norm), log_resp
+
+    def _m_step(self, X, log_resp):
+        n_samples, _ = X.shape
+        self.weights_, self.means_, self.covariances_ = (
```
Comment (unnonouno) [inline examples/gmm/gmm.py]: Use backslash

File: examples/gmm/gmm_train.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    max_iter = 30
+    repeat = 1
+
+    estimator_cpu = gmm.GaussianMixture(n_components=n_classes,
+                                        max_iter=max_iter, seed=0)
+    with timer(' CPU '):
+        for i in range(repeat):
+            train_acc, test_acc = run_gmm(X_train, y_train, X_test, y_test,
+                                          estimator_cpu, n_classes)
+    print("train_accuracy : %f" % train_acc)
```
Comment (unnonouno) [inline examples/gmm/gmm_train.py]: Use single quotes

File: examples/gmm/gmm.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
+import numpy as np
+
+import cupy
+
+
+class GaussianMixture(object):
+
+    def __init__(self, n_components=1, tol=1e-3, reg_covar=1e-6, max_iter=100,
```
Comment (unnonouno) [inline examples/gmm/gmm.py]: How about adding argument `gpu`?

File: examples/gmm/gmm.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    print('train_accuracy : %f' % train_accuracy)
+    print('test_accuracy : %f' % test_accuracy)
+    return y_test_pred, means, cov
+
+
+def draw(X, pred, means, covariances, output):
+    xp = cupy.get_array_module(X)
+    for i in six.moves.range(2):
+        labels = X[pred == i]
+        if xp == cupy:
```
Comment (niboshi) [inline examples/gmm/gmm.py]: Use `is` operator.

File: examples/gmm/gmm.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+            labels = labels.get()
+        plt.scatter(labels[:, 0], labels[:, 1], color=np.random.rand(3, 1))
+    if xp == cupy:
```
Comment (niboshi) [inline examples/gmm/gmm.py]: Ditto (use `is` operator)

File: examples/gmm/gmm.py [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+            y_test_pred, means, cov = calc_acc(X_train, y_train, X_test,
+                                               y_test, max_iter, tol)
+
+    with cupy.cuda.Device(gpuid):
+        X_train_gpu = cupy.array(X_train)
+        y_train_gpu = cupy.array(y_train)
+        y_test_gpu = cupy.array(y_test)
+        X_test_gpu = cupy.array(X_test)
+        with timer(' GPU '):
+            for i in range(repeat):
```
Comment (niboshi) [inline examples/gmm/gmm.py]: Use `six.moves.range`.

File: examples/gmm/gmm.py [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+    X, Y = np.meshgrid(x, y)
+    for i in six.moves.range(2):
+        Z = mlab.bivariate_normal(X, Y, np.sqrt(covariances[i][0]),
+                                  np.sqrt(covariances[i][1]),
+                                  means[i][0], means[i][1])
+        plt.contour(X, Y, Z)
+    plt.savefig(output + '.png')
+
+
+def run(gpuid, max_iter, tol, output):
```
Comment (niboshi) [inline examples/gmm/gmm.py]: How about printing some header line before each of trials (CPU/GPU)? Such as Running CPU... or something like that. It's helpful when an error occurred or not converged etc.

File: examples/gmm/gmm.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+    X_train = np.r_[train1, train2]
+    test1 = np.random.normal(0, [1, 2], size=(100, 2)).astype(np.float32)
+    test2 = np.random.normal(-3, [2, 1], size=(100, 2)).astype(np.float32)
+    X_test = np.r_[test1, test2]
+    y_train = np.r_[np.zeros(500000), np.ones(500000)].astype(np.int32)
+    y_test = np.r_[np.zeros(100), np.ones(100)].astype(np.int32)
+    repeat = 1
+
+    with timer(' CPU '):
+        for i in range(repeat):
```
Comment (niboshi) [inline examples/gmm/gmm.py]: Use `six.moves.range`.

**Comment by KotaroSetoyama** [comment]
I'll add some document. Please wait to merge.

**Comment by KotaroSetoyama** [comment]
Thank you for the review. I added plain document.

**Comment by niboshi** [comment]
Thank you !
@unnonouno Can you check and merge?

File: examples/gmm/gmm.py [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--gpu-id', '-g', default=0, type=int,
+                        help='ID of GPU.')
+    parser.add_argument('--max-iter', '-m', default=30, type=int,
+                        help='number of iterations')
+    parser.add_argument('--tol', '-t', default=1e-3, type=float,
+                        help='Error tolerance to stop iterations')
```
Comment (unnonouno) [inline examples/gmm/gmm.py]: Use lower case.

**Comment by KotaroSetoyama** [comment]
Thank you for the review. I've changed CPU and GPU to use same random parameters.

File: examples/gmm/gmm.py [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+    train2 = np.random.normal(-1, scale, size=(num, dim)).astype(np.float32)
+    X_train = np.r_[train1, train2]
+    test1 = np.random.normal(1, scale, size=(100, dim)).astype(np.float32)
+    test2 = np.random.normal(-1, scale, size=(100, dim)).astype(np.float32)
+    X_test = np.r_[test1, test2]
+    y_train = np.r_[np.zeros(num), np.ones(num)].astype(np.int32)
+    y_test = np.r_[np.zeros(100), np.ones(100)].astype(np.int32)
+
+    repeat = 5
+    for i in six.moves.range(repeat):
```
Comment (unnonouno) [inline examples/gmm/gmm.py]: Don't repeat. It is not required.

**Comment by unnonouno** [comment]
OK, LGTM

**Pull request merged**